README.md
# tus-ruby-server

A Ruby server for the [tus resumable upload protocol]. It implements the core
1.0 protocol, along with the following extensions:

* [`creation`][creation] (and `creation-defer-length`)
* [`concatenation`][concatenation]
* [`checksum`][checksum]
* [`expiration`][expiration]
* [`termination`][termination]

## Installation

```rb
# Gemfile
gem "tus-server", "~> 2.3"
```

## Usage

The gem provides a `Tus::Server` [Roda] app, which you can mount inside your
main application. If you're using Rails, you can mount it in `config/routes.rb`:

```rb
# config/routes.rb (Rails)
Rails.application.routes.draw do
  mount Tus::Server => "/files"
end
```

Now you can tell your tus client library (e.g. [tus-js-client]) to use this
endpoint:

```js
// using tus-js-client
new tus.Upload(file, {
  endpoint: '/files',
  chunkSize: 5*1024*1024, // chunking is required unless you're using Falcon
  // ...
})

// OR

// using Uppy
uppy.use(Uppy.Tus, {
  endpoint: '/files',
  chunkSize: 5*1024*1024, // chunking is required unless you're using Falcon
})
```

By default uploaded files will be stored in the `data/` directory. After the
upload is complete, you'll probably want to attach the uploaded file to a
database record. [Shrine] is a file attachment library that provides
integration with tus-ruby-server, see [this walkthrough][shrine resumable
walkthrough] that adds resumable uploads from scratch, and for a complete
example you can check out the [demo app][shrine-tus-demo].

### Streaming web server

Running the tus server alongside your main app using popular web servers like
Puma or Unicorn is probably fine for most cases, but it does come with a few
gotchas. First, since these web servers don't accept partial requests (request
where the request body hasn't been fully received), the tus client must be
configured to split the upload into multiple requests. Second, since web
workers are tied for the duration of the request, serving uploaded files
through the tus server app could significantly impact request throughput; this
can be avoided by having your frontend server (Nginx) serve the files if using
`Filesystem` storage, or if you're using a cloud service like S3 having
download requests redirect to the service file URL.

That being said, there is a ruby web server that addresses these limitations
– [Falcon]. Falcon is part of the [`async` ecosystem][async], and it utilizes
non-blocking IO to process requests and responses in a streaming fashion
without tying up your web workers. This has several benefits for
`tus-ruby-server`:

* since tus server is called to handle the request as soon as the request
  headers are received, data from the request body will be uploaded to the
  configured storage as it's coming in from the client

* your web workers don't get tied up waiting for the client data, because as
  soon as the server needs to wait for more data from the client, Falcon's
  reactor switches to processing another request

* if the upload request that's in progress gets interrupted, tus server will be
  able save data that has been received so far, so it's not necessary for the
  tus client to split the upload into multiple chunks

* when uploaded files are being downloaded from the tus server, the request
  throughput won't be impacted by the speed in which the client retrieves the
  response body, because Falcon's reactor will switch to another request if the
  client buffer gets full

Falcon provides a Rack adapter and is compatible with Rails, so you can use it
even if you're mounting `tus-ruby-server` inside your main app, just add
`falcon` to the Gemfile and run `falcon serve`.

```rb
# Gemfile
gem "falcon"
```
```sh
$ falcon serve # reads from config.ru and starts the server
```

Alternatively, you can run `tus-ruby-server` as a standalone app, by creating
a separate "rackup" file just for it and pointing Falcon to that rackup file:

```rb
# tus.ru
require "tus/server"

# ... configure tus-ruby-server ...

map "/files" do
  run Tus::Server
end
```
```sh
$ falcon serve --config tus.ru
```

## Storage

### Filesystem

By default `Tus::Server` stores uploaded files in the `data/` directory. You
can configure a different directory:

```rb
require "tus/storage/filesystem"

Tus::Server.opts[:storage] = Tus::Storage::Filesystem.new("public/tus")
```

If the configured directory doesn't exist, it will automatically be created.
By default the UNIX permissions applied will be 0644 for files and 0755 for
directories, but you can set different permissions:

```rb
Tus::Storage::Filesystem.new("data", permissions: 0600, directory_permissions: 0777)
```

One downside of filesystem storage is that it doesn't work by default if you
want to run tus-ruby-server on multiple servers, you'd have to set up a shared
filesystem between the servers. Another downside is that you have to make sure
your servers have enough disk space. Also, if you're using Heroku, you cannot
store files on the filesystem as they won't persist.

All these are reasons why you might store uploaded data on a different storage,
and luckily tus-ruby-server ships with two more storages.

#### Serving files

If your retrieving uploaded files through the download endpoint, by default the
files will be served through the Ruby application. However, that's very
inefficient, as web workers are tied when serving download requests and cannot
serve additional requests for that duration.

Therefore, it's highly recommended to delegate serving uploaded files to your
frontend server (Nginx, Apache). This can be achieved with the
`Rack::Sendfile` middleware, see its [documentation][Rack::Sendfile] to learn
more about how to use it with popular frontend servers.

If you're using Rails, you can enable the `Rack::Sendfile` middleware by
setting the `config.action_dispatch.x_sendfile_header` value accordingly:

```rb
Rails.application.config.action_dispatch.x_sendfile_header = "X-Sendfile" # Apache and lighttpd
# or
Rails.application.config.action_dispatch.x_sendfile_header = "X-Accel-Redirect" # Nginx
```

Otherwise you can add the `Rack::Sendfile` middleware to the stack in
`config.ru`:

```rb
use Rack::Sendfile, "X-Sendfile" # Apache and lighttpd
# or
use Rack::Sendfile, "X-Accel-Redirect" # Nginx
```

### Amazon S3

You can switch to `Tus::Storage::S3` to uploads files to AWS S3 using the
multipart API. For this you'll also need to add the [aws-sdk-s3] gem to your
Gemfile.

```rb
# Gemfile
gem "aws-sdk-s3", "~> 1.2"
```

```rb
require "tus/storage/s3"

# You can omit AWS credentials if you're authenticating in other ways
Tus::Server.opts[:storage] = Tus::Storage::S3.new(
  bucket:            "my-app", # required
  access_key_id:     "abc",
  secret_access_key: "xyz",
  region:            "eu-west-1",
)
```

If you want to files to be stored in a certain subdirectory, you can specify
a `:prefix` in the storage configuration.

```rb
Tus::Storage::S3.new(prefix: "tus", **options)
```

You can also specify additional options that will be fowarded to
[`Aws::S3::Client#create_multipart_upload`] using `:upload_options`.

```rb
Tus::Storage::S3.new(upload_options: { acl: "public-read" }, **options)
```

All other options will be forwarded to [`Aws::S3::Client#initialize`]:

```rb
Tus::Storage::S3.new(
  use_accelerate_endpoint: true,
  logger: Logger.new(STDOUT),
  retry_limit: 5,
  http_open_timeout: 10,
  # ...
)
```

If you're using [concatenation], you can specify the concurrency in which S3
storage will copy partial uploads to the final upload (defaults to `10`):

```rb
Tus::Storage::S3.new(concurrency: { concatenation: 20 }, **options)
```

#### Limits

Be aware that the AWS S3 Multipart Upload API has the following limits:

| Item                               | Specification                         |
| ----                               | -------------                         |
| Part size                          | 5 MB to 5 GB, last part can be < 5 MB |
| Maximum number of parts per upload | 10,000                                |
| Maximum object size                | 5 TB                                  |

This means that if you're chunking uploads in your tus client, the chunk size
needs to be **5 MB or larger**. Furthermore, if you're allowing your users to
upload files larger than 50 GB, you will the minimum chunk size needs to be
higher (`ceil(max_length, max_multipart_parts)`). Note that chunking is
optional if you're running on Falcon, but it's mandatory if you're using Puma
or another web server.

`Tus::Storage::S3` is relying on the above limits for determining the multipart
part size. If you're using a different S3-compatible service which has different
limits, you should pass them in when initializing the storage:

```rb
Tus::Storage::S3.new(limits: {
  min_part_size:       5 * 1024 * 1024,
  max_part_size:       5 * 1024 * 1024 * 1024,
  max_multipart_parts: 10_000,
  max_object_size:     5 * 1024 * 1024 * 1024,
}, **options)
```

#### Serving files

If you'll be retrieving uploaded files through the tus server app, it's
recommended to set `Tus::Server.opts[:redirect_download]` to `true`. This will
avoid tus server downloading and serving the file from S3, and instead have the
download endpoint redirect to the direct S3 object URL.

```rb
Tus::Server.opts[:redirect_download] = true
```

You can customize how the S3 object URL is being generated by passing a block
to `:redirect_download`, which will then be evaluated in the context of the
`Tus::Server` instance (which allows accessing the `request` object). See
[`Aws::S3::Object#get`] for the list of options that
`Tus::Storage::S3#file_url` accepts.

```rb
Tus::Server.opts[:redirect_download] = -> (uid, info, **options) do
  storage.file_url(uid, info, expires_in: 10, **options) # link expires after 10 seconds
end
```

### Google Cloud Storage, Microsoft Azure Blob Storage

While tus-ruby-server doesn't currently ship with integrations for Google Cloud
Storage or Microsoft Azure Blob Storage, you can still use these services via
**[Minio]**.

Minio is an open source distributed object storage server that implements the
Amazon S3 API and provides gateways for [GCP][minio gcp] and [Azure][minio
azure]. This means it's possible to use the existing S3 tus-ruby-server
integration to point to the Minio server, which will then in turn forward those
calls to GCP or Azure.

Let's begin by installing Minio via Homebrew:

```sh
$ brew install minio/stable/minio
```

And starting the Minio server as a gateway to GCP or Azure:

```sh
# Google Cloud Storage
$ export GOOGLE_APPLICATION_CREDENTIALS=/path/credentials.json
$ export MINIO_ACCESS_KEY=minioaccesskey
$ export MINIO_SECRET_KEY=miniosecretkey
$ minio gateway gcs yourprojectid
```
```sh
# Microsoft Azure Blob Storage
$ export MINIO_ACCESS_KEY=azureaccountname
$ export MINIO_SECRET_KEY=azureaccountkey
$ minio gateway azure
```

Now follow the displayed link to open the Minio web UI and create a bucket in
which you want your files to be stored. Once you've done that, you can
initialize `Tus::Storage::S3` to point to your Minio server and bucket:

```rb
Tus::Storage::S3.new(
  access_key_id:     "<MINIO_ACCESS_KEY>", # "AccessKey" value
  secret_access_key: "<MINIO_SECRET_KEY>", # "SecretKey" value
  endpoint:          "<MINIO_ENDPOINT>",   # "Endpoint"  value
  bucket:            "<MINIO_BUCKET>",     # name of the bucket you created
  region:            "us-east-1",
  force_path_style:  true,
)
```

### MongoDB GridFS

MongoDB has a specification for storing and retrieving large files, called
"[GridFS]". Tus-ruby-server ships with `Tus::Storage::Gridfs` that you can
use, which uses the [Mongo] gem.

```rb
# Gemfile
gem "mongo", "~> 2.3"
```

```rb
require "tus/storage/gridfs"

client = Mongo::Client.new("mongodb://127.0.0.1:27017/mydb")
Tus::Server.opts[:storage] = Tus::Storage::Gridfs.new(client: client)
```

You can change the database prefix (defaults to `fs`):

```rb
Tus::Storage::Gridfs.new(client: client, prefix: "fs_temp")
```

By default MongoDB Gridfs stores files in chunks of 256KB, but you can change
that with the `:chunk_size` option:

```rb
Tus::Storage::Gridfs.new(client: client, chunk_size: 1*1024*1024) # 1 MB
```

Note that if you're using the [concatenation] tus feature with Gridfs, all
partial uploads except the last one are required to fill in their Gridfs
chunks, meaning the length of each partial upload needs to be a multiple of the
`:chunk_size` number.

### Other storages

If none of these storages suit you, you can write your own, you just need to
implement the same public interface:

```rb
def create_file(uid, info = {})            ... end
def concatenate(uid, part_uids, info = {}) ... end
def patch_file(uid, io, info = {})         ... end
def update_info(uid, info)                 ... end
def read_info(uid)                         ... end
def get_file(uid, info = {}, range: nil)   ... end
def file_url(uid, info = {}, **options)    ... end # optional
def delete_file(uid, info = {})            ... end
def expire_files(expiration_date)          ... end
```

## Hooks

You can register code to be executed on the following events:

* `before_create` – before upload has been created
* `after_create` – before upload has been created
* `after_finish` – after the last chunk has been stored
* `after_terminate` – after the upload has been deleted

Each hook also receives two parameters: ID of the upload (`String`) and
additional information about the upload (`Tus::Info`).

```rb
Tus::Server.after_finish do |uid, info|
  uid  #=> "c0b67b04a9eccb4b1202000de628964f"
  info #=> #<Tus::Info>

  info.length        #=> 10      (Upload-Length)
  info.offset        #=> 0       (Upload-Offset)
  info.metadata      #=> {...}   (Upload-Metadata)
  info.expires       #=> #<Time> (Upload-Expires)

  info.partial?      #=> false   (Upload-Concat)
  info.final?        #=> false   (Upload-Concat)

  info.defer_length? #=> false   (Upload-Defer-Length)
end
```

Because each hook is evaluated inside the `Tus::Server` instance (which is also
a [Roda] instance), you can access request information and set response status
and headers:

```rb
Tus::Server.after_terminate do |uid, info|
  self     #=> #<Tus::Server> (and #<Roda>)
  request  #=> #<Roda::Request>
  response #=> #<Roda::Response>
end
```

So, with hooks you could for example add authentication to `Tus::Server`:

```rb
Tus::Server.before_create do |uid, info|
  authenticated = Authentication.call(request.headers["Authorization"])

  unless authenticated
    response.status = 403 # Forbidden
    response.write("Not authenticated")
    request.halt # return response now
  end
end
```

If you want to add hooks on more types of events, you can use Roda's
[hooks][roda hooks] plugin to set `before` or `after` hooks for any request,
which are also evaluated in context of the `Roda` instance:

```rb
Tus::Server.plugin :hooks
Tus::Server.before do
  # called before each Roda request
end
Tus::Server.after do
  # called after each Roda request
end
```

Provided that you're mounting `Tus::Server` inside your main app, any Rack
middlewares that your main app uses will also be called for requests routed to
the `Tus::Server`. If you want to add Rack middlewares to `Tus::Server`, you
can do it by calling `Tus::Server.use`:

```rb
Tus::Server.use SomeRackMiddleware
```

## Maximum size

By default the size of files the tus server will accept is unlimited, but you
can configure the maximum file size:

```rb
Tus::Server.opts[:max_size] = 5 * 1024*1024*1024 # 5GB
```

## Expiration

Tus-ruby-server automatically adds expiration dates to each uploaded file, and
refreshes this date on each PATCH request. By default files expire 7 days after
they were last updated, but you can modify `:expiration_time`:

```rb
Tus::Server.opts[:expiration_time] = 2*24*60*60 # 2 days
```

Tus-ruby-server won't automatically delete expired files, but each storage
knows how to expire old files, so you just have to set up a recurring task
that will call `#expire_files`.

```rb
expiration_time = Tus::Server.opts[:expiration_time]

CHANGELOG.md
## HEAD

* Deprecate `:concurrency` option in S3 storage (@janko)

* Don't use a thread pool in `S3#concatenate` (@janko)

## 2.3.0 (2019-05-14)

* Allow uploading files larger than 50 GB to S3 storage by scaling the part size according to upload length (@janko)

* Allow S3 limits to be overriden via the `:limits` option (@janko)

* Reject uploads larger than 5 TB in `S3#create_file` (@janko)

* Remove superfluous `#head_object` operation from `S3#concatenate` (@janko)

* Validate that partial uploads are finished on the concatenation request (@janko)

* Unparallelize validating partial uploads for concatentation (@janko)

* Move total length calculation for concatenation into `Tus::Server` (@janko)

* Allow `Upload-Concat` and `Upload-Defer-Length` headers in CORS headers (@janko)

* Always override `:content_disposition` option with the value from `:upload_options` in `S3#create_file` (@janko)

* Apply `name`/`type` metadata to `:content_type`/`:content_disposition` options in `S3#create_file` (@janko)

* Use `content_disposition` gem for `:content_disposition` option in `S3#create_file` (@janko)

* Remove checks whether multipart upload was successfully aborted in S3 storage (@janko)

* Don't return any `Content-Type` when type is not specified in metadata (@janko)

* Add `ETag` header to download endpoint to prevent `Rack::ETag` buffering file content (@janko)

* Take `:prefix` into account in `Tus::Storage::S3#expire_files` (@janko)

## 2.2.1 (2018-12-19)

* Use `content_disposition` gem to generate `Content-Disposition` in download endpoint (@janko)

## 2.2.0 (2018-12-02)

* Add `before_create`, `after_create`, `after_finish`, and `after_terminate` hooks (@janko)

* Rename `Tus::Info#concatenation?` to `Tus::Info#final?` (@janko)

* Use `Storage#concurrency` for parallelized retrieval of partial uploads in `Upload-Concat` validation (@janko)

* Replace `:thread_count` with `:concurrency` in S3 storage (@janko)

* Validate that sum of partial uploads doesn't exceed `Tus-Max-Size` on concatenation (@janko)

* Drop MRI 2.2 support (@janko)

* Accept absolute URLs of partial uploads when creating a final upload (@janko)

## 2.1.2 (2018-10-21)

* Make tus-ruby-server fully work with non-rewindable Rack input (@janko)

## 2.1.1 (2018-05-26)

* Rename `:download_url` option to `:redirect_download` (@janko)

## 2.1.0 (2018-05-15)

* Add `:download_url` server option for redirecting to a download URL (@janko)

* Allow application servers to serve files stored on disk via the `Rack::Sendfile` middleware (@janko)

* Reject `Upload-Metadata` which contains key-value pairs separated by spaces (@janko)

* Don't overwite info file if it already exists in `Tus::Storage::FileSystem` (@janko)

## 2.0.2 (2017-12-24)

* Handle `name` and `type` metadata for Uppy compatibility (@janko)

## 2.0.1 (2017-11-13)

* Add back support for Roda 2.x (@janko)

## 2.0.0 (2017-11-13)

* Upgrade to Roda 3 (@janko)

* Remove deprecated support for aws-sdk 2.x in `Tus::Storage::S3` (@janko)

* Drop official support for MRI 2.1 (@janko)

* Add generic `Tus::Response` class that storages can use (@janko)

* Remove `Tus::Response#length` (@janko)

* Remove deprecated Goliath integration (@janko)

* Return `400 Bad Request` instead of `404 Not Found` when some partial uploads are missing in a concatenation request (@janko)

* Use Rack directly instead of Roda's `streaming` plugin for downloding (@janko)

## 1.2.1 (2017-11-05)

* Improve communication when handling `aws-sdk 2.x` fallback in `Tus::Storage::S3` (@janko)

## 1.2.0 (2017-09-18)

* Deprecate `aws-sdk` 2.x in favour of the new `aws-sdk-s3` gem (@janko)

## 1.1.3 (2017-09-17)

* Return `Accept-Ranges: bytes` response header in download endpoint (@janko)

## 1.1.2 (2017-09-12)

* Add support for the new `aws-sdk-s3` gem (@janko)

## 1.1.1 (2017-07-23)

* Restore backwards compatibility with MRI 2.1 and MRI 2.2 that was broken in previous release (@janko)

## 1.1.0 (2017-07-23)

* Ignore retryable networking errors in `Tus::Storage::S3#patch_file` for resiliency (@janko)

* Deprecate `Tus::Server::Goliath` in favour of [goliath-rack_proxy](https://github.com/janko/goliath-rack_proxy) (@janko)

* Reduce string allocations in MRI 2.3+ with `frozen-string-literal: true` magic comments (@janko)

## 1.0.0 (2017-07-17)

* Add Goliath integration (@janko)

* [BREAKING] Save data in `"#{uid}"` instead of `"#{uid}.file"` in `Tus::Storage::Filesystem` (@janko)

* Modify S3 storage to cache chunks into memory instead of disk, which reduces disk IO (@janko)

* [BREAKING] Require each storage to return the number of bytes uploaded in `#patch_file` (@janko)

* Make S3 storage upload all received data from `tus-js-client` that doesn't have max chunk size configured (@janko)

* Verify that all partial uploads have `Upload-Concat: partial` before concatenation (@janko)

* Include CORS and tus response headers in 404 responses (@janko)

* Improve streaming on dynamic Rack inputs such as `Unicorn::TeeInput` for S3 and Gridfs storage (@janko)

* Terminate HTTP connection to S3 when response is closed (@janko)

* Allow `Transfer-Encoding: chunked` to be used, meaning `Content-Length` can be blank (@janko)

* Remove newlines from the base64-encoded CRC32 signature (@janko)

* Lazily require `digest`, `zlib`, and `base64` standard libraries (@janko)

## 0.10.2 (2017-04-19)

* Allow empty metadata values in `Upload-Metadata` header (@lysenkooo)

## 0.10.1 (2017-04-13)

* Fix download endpoint returning incorrect response body in some cases in development (@janko)

* Remove `concatenation-unfinished` from list of supported extensions (@janko)

## 0.10.0 (2017-03-27)

* Fix invalid `Content-Disposition` header in GET requests to due mutation of `Tus::Server.opts[:disposition]` (@janko)

* Make `Response` object from `Tus::Server::S3` also respond to `#close` (@janko)

* Don't return `Content-Type` header when there is no content returned (@janko)


lib/tus/server.rb
# frozen-string-literal: true

require "roda"
require "rack/rewindable_input"

require "tus/storage/filesystem"
require "tus/info"
require "tus/input"
require "tus/checksum"
require "tus/errors"

require "content_disposition"

require "securerandom"
require "time"

module Tus
  class Server < Roda
    SUPPORTED_VERSIONS = ["1.0.0"]
    SUPPORTED_EXTENSIONS = [
      "creation", "creation-defer-length",
      "termination",
      "expiration",
      "concatenation",
      "checksum",
    ]
    SUPPORTED_CHECKSUM_ALGORITHMS = %w[sha1 sha256 sha384 sha512 md5 crc32]
    RESUMABLE_CONTENT_TYPE = "application/offset+octet-stream"
    HOOKS = %i[before_create after_create after_finish after_terminate]

    opts[:max_size]          = nil
    opts[:expiration_time]   = 7*24*60*60
    opts[:disposition]       = "inline"
    opts[:redirect_download] = nil
    opts[:hooks]             = {}

    plugin :all_verbs
    plugin :default_headers, {"Content-Type" => ""}
    plugin :delete_empty_headers
    plugin :request_headers
    plugin :not_allowed

    HOOKS.each do |hook|
      define_singleton_method(hook) do |&block|
        opts[:hooks][hook] = block
      end

      define_method(hook) do |*args|
        instance_exec(*args, &opts[:hooks][hook]) if opts[:hooks][hook]
      end
    end

    route do |r|
      if request.headers["X-HTTP-Method-Override"]
        request.env["REQUEST_METHOD"] = request.headers["X-HTTP-Method-Override"]
      end

      response.headers.update(
        "Tus-Resumable" => SUPPORTED_VERSIONS.first,
      )

      handle_cors!
      validate_tus_resumable! unless request.options? || request.get?

      r.is ['', true] do
        # OPTIONS /
        r.options do
          response.headers.update(
            "Tus-Version"            => SUPPORTED_VERSIONS.join(","),
            "Tus-Extension"          => SUPPORTED_EXTENSIONS.join(","),
            "Tus-Max-Size"           => max_size.to_s,
            "Tus-Checksum-Algorithm" => SUPPORTED_CHECKSUM_ALGORITHMS.join(","),
          )

          no_content!
        end

        # POST /
        r.post do
          validate_upload_length! unless request.headers["Upload-Concat"].to_s.start_with?("final") || request.headers["Upload-Defer-Length"] == "1"
          validate_upload_metadata! if request.headers["Upload-Metadata"]
          validate_upload_concat! if request.headers["Upload-Concat"]

          uid  = SecureRandom.hex
          info = Tus::Info.new(
            "Upload-Length"       => request.headers["Upload-Length"],
            "Upload-Offset"       => "0",
            "Upload-Defer-Length" => request.headers["Upload-Defer-Length"],
            "Upload-Metadata"     => request.headers["Upload-Metadata"],
            "Upload-Concat"       => request.headers["Upload-Concat"],
            "Upload-Expires"      => (Time.now + expiration_time).httpdate,
          )

          before_create(uid, info)

          if info.final?
            length = validate_partial_uploads!(info.partial_uploads)

            storage.concatenate(uid, info.partial_uploads, info.to_h)
            info["Upload-Length"] = length.to_s
            info["Upload-Offset"] = length.to_s
          else
            storage.create_file(uid, info.to_h)
          end

          after_create(uid, info)

          storage.update_info(uid, info.to_h)
          response.headers.update(info.headers)

          file_url = "#{request.url.chomp("/")}/#{uid}"
          created!(file_url)
        end
      end

      r.is String do |uid|
        # OPTIONS /{uid}
        r.options do
          response.headers.update(
            "Tus-Version"            => SUPPORTED_VERSIONS.join(","),
            "Tus-Extension"          => SUPPORTED_EXTENSIONS.join(","),
            "Tus-Max-Size"           => max_size.to_s,
            "Tus-Checksum-Algorithm" => SUPPORTED_CHECKSUM_ALGORITHMS.join(","),
          )

          no_content!
        end

        begin
          info = Tus::Info.new(storage.read_info(uid))
        rescue Tus::NotFound
          error!(404, "Upload Not Found")
        end

        # HEAD /{uid}
        r.head do
          response.headers.update(info.headers)
          response.headers["Cache-Control"] = "no-store"

          no_content!
        end

        # PATCH /{uid}
        r.patch do
          if info.defer_length? && request.headers["Upload-Length"]
            validate_upload_length!

            info["Upload-Length"]       = request.headers["Upload-Length"]
            info["Upload-Defer-Length"] = nil
          end

          input = get_input(info)

          validate_content_type!
          validate_upload_offset!(info)
          validate_content_length!(request.content_length.to_i, info) if request.content_length

          begin
            if request.headers["Upload-Checksum"]
              input = Rack::RewindableInput.new(input) unless input.rewindable?
              validate_upload_checksum!(input)
            end

            bytes_uploaded = storage.patch_file(uid, input, info.to_h)
          rescue Tus::MaxSizeExceeded
            validate_content_length!(input.pos, info)
          end

          info["Upload-Offset"] = (info.offset + bytes_uploaded).to_s
          info["Upload-Expires"] = (Time.now + expiration_time).httpdate

          if info.offset == info.length # last chunk
            storage.finalize_file(uid, info.to_h) if storage.respond_to?(:finalize_file)

            after_finish(uid, info)
          end

          storage.update_info(uid, info.to_h)
          response.headers.update(info.headers)

          no_content!
        end

        # GET /{uid}
        r.get do
          validate_upload_finished!(info)

          if redirect_download
            redirect_url = instance_exec(uid, info.to_h,
              content_type:        info.type,
              content_disposition: ContentDisposition.(disposition: opts[:disposition], filename: info.name),
              &redirect_download)

            r.redirect redirect_url
          else
            range = handle_range_request!(info.length)

            response.headers["Content-Disposition"] = ContentDisposition.(disposition: opts[:disposition], filename: info.name)
            response.headers["Content-Type"]        = info.type if info.type
            response.headers["ETag"]                = %(W/"#{uid}")

            body = storage.get_file(uid, info.to_h, range: range)

            r.halt response.finish_with_body(body)
          end
        end

        # DELETE /{uid}
        r.delete do
          storage.delete_file(uid, info.to_h)

          after_terminate(uid, info)

          no_content!
        end
      end
    end

    # Wraps the Rack input (request body) into a Tus::Input object, applying a
    # size limit if one exists.
    def get_input(info)
      offset = info.offset
      total  = info.length || max_size
      limit  = total - offset if total

      Tus::Input.new(request.body, limit: limit)
    end

    def validate_content_type!
      error!(415, "Invalid Content-Type header") if request.content_type != RESUMABLE_CONTENT_TYPE
    end

    def validate_tus_resumable!
      client_version = request.headers["Tus-Resumable"]

      unless SUPPORTED_VERSIONS.include?(client_version)
        response.headers["Tus-Version"] = SUPPORTED_VERSIONS.join(",")
        error!(412, "Unsupported version")
      end
    end

    def validate_upload_length!
      upload_length = request.headers["Upload-Length"]

      error!(400, "Missing Upload-Length header") if upload_length.to_s == ""
      error!(400, "Invalid Upload-Length header") if upload_length =~ /\D/
      error!(400, "Invalid Upload-Length header") if upload_length.to_i < 0

      if max_size && upload_length.to_i > max_size
        error!(413, "Upload-Length header too large")
      end
    end

    def validate_upload_offset!(info)
      upload_offset = request.headers["Upload-Offset"]

      error!(400, "Missing Upload-Offset header") if upload_offset.to_s == ""
      error!(400, "Invalid Upload-Offset header") if upload_offset =~ /\D/
      error!(400, "Invalid Upload-Offset header") if upload_offset.to_i < 0

      if upload_offset.to_i != info.offset
        error!(409, "Upload-Offset header doesn't match current offset")
      end
    end

    def validate_content_length!(size, info)
      if info.length
        error!(403, "Cannot modify completed upload") if info.offset == info.length
        error!(413, "Size of this chunk surpasses Upload-Length") if info.offset + size > info.length
      elsif max_size
        error!(413, "Size of this chunk surpasses Tus-Max-Size") if info.offset + size > max_size
      end
    end

    def validate_upload_finished!(info)
      error!(403, "Cannot download unfinished upload") unless info.length == info.offset
    end

    def validate_upload_metadata!
      upload_metadata = request.headers["Upload-Metadata"]

      upload_metadata.split(",").each do |string|
        key, value = string.split(" ", 2)

        error!(400, "Invalid Upload-Metadata header") if key.nil?
        error!(400, "Invalid Upload-Metadata header") if key.ord > 127
        error!(400, "Invalid Upload-Metadata header") if key =~ /,| /

        error!(400, "Invalid Upload-Metadata header") if value =~ /[^a-zA-Z0-9+\/=]/
      end
    end

    def validate_upload_concat!
      upload_concat = request.headers["Upload-Concat"]

      error!(400, "Invalid Upload-Concat header") if upload_concat !~ /^(partial|final)/

      if upload_concat.start_with?("final")
        string = upload_concat.split(";").last
        string.split(" ").each do |url|
          error!(400, "Invalid Upload-Concat header") if url !~ /#{request.script_name}\/\w+$/
        end
      end
    end

    # Validates that each partial upload exists and is marked as one, and at the
    # same time calculates the sum of part lengths.
    def validate_partial_uploads!(part_uids)
      length = 0

      part_uids.each do |part_uid|
        begin
          part_info = storage.read_info(part_uid)
        rescue Tus::NotFound
          error!(400, "Partial upload not found")
        end

        part_info = Tus::Info.new(part_info)

        error!(400, "Upload is not partial") unless part_info.partial?

        unless part_info.length == part_info.offset
          error!(400, "Partial upload is not finished")
        end

        length += part_info.length
      end

      if max_size && length > max_size
        error!(400, "The sum of partial upload lengths exceeds Tus-Max-Size")
      end

      length
    end

    def validate_upload_checksum!(input)
      algorithm, checksum = request.headers["Upload-Checksum"].split(" ")

      error!(400, "Invalid Upload-Checksum header") if algorithm.nil? || checksum.nil?
      error!(400, "Invalid Upload-Checksum header") unless SUPPORTED_CHECKSUM_ALGORITHMS.include?(algorithm)

      generated_checksum = Tus::Checksum.generate(algorithm, input)
      error!(460, "Upload-Checksum value doesn't match generated checksum") if generated_checksum != checksum
    end

    # Handles partial responses requested in the "Range" header. Implementation
    # is mostly copied from Rack::File.
    def handle_range_request!(length)
      if Rack.release >= "2.0"
        ranges = Rack::Utils.get_byte_ranges(request.headers["Range"], length)
      else
        ranges = Rack::Utils.byte_ranges(request.env, length)
      end

      # we support ranged requests
      response.headers["Accept-Ranges"] = "bytes"

      if ranges.nil? || ranges.length > 1
        # no ranges, or multiple ranges (which we don't support)
        response.status = 200
        range = 0..length-1
      elsif ranges.empty?
        # unsatisfiable range
        response.headers["Content-Range"] = "bytes */#{length}"
        error!(416, "Byte range unsatisfiable")
      else
        range = ranges[0]
        response.status = 206
        response.headers["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{length}"
      end

      response.headers["Content-Length"] = range.size.to_s

      range
    end

    def redirect_download
      value = opts[:redirect_download]

      if opts[:download_url]
        value ||= opts[:download_url]
        warn "[TUS-RUBY-SERVER DEPRECATION] The :download_url option has been renamed to :redirect_download."
      end

      value = storage.method(:file_url) if value == true

      value
    end

    def handle_cors!
      origin = request.headers["Origin"]

      return if origin.to_s == ""

      response.headers["Access-Control-Allow-Origin"] = origin

      if request.options?
        response.headers["Access-Control-Allow-Methods"] = "POST, GET, HEAD, PATCH, DELETE, OPTIONS"
        response.headers["Access-Control-Allow-Headers"] = "Origin, X-Requested-With, Content-Type, Upload-Length, Upload-Offset, Tus-Resumable, Upload-Metadata, Upload-Defer-Length, Upload-Concat"
        response.headers["Access-Control-Max-Age"]       = "86400"
      else
        response.headers["Access-Control-Expose-Headers"] = "Upload-Offset, Location, Upload-Length, Tus-Version, Tus-Resumable, Tus-Max-Size, Tus-Extension, Upload-Metadata, Upload-Defer-Length, Upload-Concat"
      end
    end

    def no_content!
      response.status = 204
      request.halt
    end

    def created!(location)
      response.status = 201
      response.headers["Location"] = location
      request.halt
    end

    def error!(status, message)
      response.status = status
      response.write(message) unless request.head?
      response.headers["Content-Type"] = "text/plain"
      request.halt
    end

    def storage
      opts[:storage] || Tus::Storage::Filesystem.new("data")
    end

    def max_size
      opts[:max_size]
    end

    def expiration_time
      opts[:expiration_time]
    end

    def expiration_interval
      opts[:expiration_interval]
    end
  end
end
